File: scripts/build_csv.mjs
// Deterministic CSV builder (Node fallback) (shared core) for the estimator plugin.
import { readFileSync, writeFileSync, existsSync, mkdirSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { pathToFileURL } from "node:url";

export const HEADERS = [
  "Key (key)",
  "Parent (parent_key)",
  "Issue Type (issue_type)",
  "Title (summary)",
  "Description (description)",
  "Components (components)",
  "Labels (labels)",
  "Story Points (story_points)",
  "Time Spent [s] (time_spent_seconds)",
  "Resolution Date (resolution_date)",
  "Comments (comments)",
  "Design Link (design_link)",
  "Estimate Basis (estimate_basis)",
];

export const FIELD_KEYS = HEADERS.map((h) => h.match(/\(([^)]+)\)\s*$/)[1]);

export const MULTIVALUE_SEP = ", ";
export const COMMENT_SEP = "\n\n";

const s = (v) => (v === null || v === undefined ? "" : String(v));
const join = (v, sep) => {
  if (v === null || v === undefined || v === "" || (Array.isArray(v) && v.length === 0)) {
    return "";
  }
  return Array.isArray(v) ? v.map(String).join(sep) : String(v);
};

export function recordToRow(rec) {
  return [
    s(rec.key),
    s(rec.parent_key),
    s(rec.issue_type),
    s(rec.summary),
    s(rec.description),
    join(rec.components, MULTIVALUE_SEP),
    join(rec.labels, MULTIVALUE_SEP),
    s(rec.story_points),
    s(rec.time_spent_seconds),
    s(rec.resolution_date),
    join(rec.comments, COMMENT_SEP),
    s(rec.design_link),
    s(rec.estimate_basis),
  ];
}

export function formatField(value) {
  const v = value === null || value === undefined ? "" : String(value);
  return /[;"\n\r]/.test(v) ? '"' + v.replace(/"/g, '""') + '"' : v;
}

export function formatRow(fields) {
  return fields.map(formatField).join(";");
}

export function parseCsv(text) {
  const rows = [];
  let row = [];
  let field = "";
  let inQuotes = false;
  let i = 0;
  while (i < text.length) {
    const ch = text[i];
    if (inQuotes) {
      if (ch === '"') {
        if (text[i + 1] === '"') { field += '"'; i += 2; continue; }
        inQuotes = false; i += 1; continue;
      }
      field += ch; i += 1; continue;
    }
    if (ch === '"') { inQuotes = true; i += 1; continue; }
    if (ch === ";") { row.push(field); field = ""; i += 1; continue; }
    if (ch === "\n" || ch === "\r") {
      if (ch === "\r" && text[i + 1] === "\n") i += 1;
      row.push(field); field = "";
      rows.push(row); row = [];
      i += 1; continue;
    }
    field += ch; i += 1;
  }
  if (field.length > 0 || row.length > 0) {
    row.push(field);
    rows.push(row);
  }
  return rows;
}

export function buildRows(existingRows, records, mode) {
  if (mode !== "overwrite" && mode !== "append") {
    throw new Error(`mode must be 'overwrite' or 'append', got: ${mode}`);
  }
  const merged = new Map();
  if (mode === "append" && existingRows) {
    for (const row of existingRows) {
      if (row && row.length) merged.set(row[0], row);
    }
  }
  for (const rec of records) {
    const row = recordToRow(rec);
    merged.set(row[0], row);
  }
  return [...merged.values()];
}

function normalize(row) {
  // Pad legacy rows that predate a column addition out to the header width.
  return row.length < HEADERS.length
    ? row.concat(Array(HEADERS.length - row.length).fill(""))
    : row;
}

export function readExisting(path) {
  if (!existsSync(path)) return [];
  const rows = parseCsv(readFileSync(path, "utf8"));
  return rows.length ? rows.slice(1).map(normalize) : [];
}

export function writeCsv(path, rows) {
  mkdirSync(dirname(resolve(path)), { recursive: true });
  const lines = [formatRow(HEADERS), ...rows.map(formatRow)];
  writeFileSync(path, lines.join("\n") + "\n", "utf8");
}

export function main(argv) {
  if (argv.length !== 3) {
    console.error("Usage: build_csv.mjs <input.json> <output.csv> <overwrite|append>");
    process.exit(1);
  }
  const [inputJson, outputCsv, mode] = argv;
  const records = JSON.parse(readFileSync(inputJson, "utf8"));
  const existing = mode === "append" ? readExisting(outputCsv) : [];
  const rows = buildRows(existing, records, mode);
  writeCsv(outputCsv, rows);
  console.log(`Wrote ${rows.length} rows to ${outputCsv}`);
  console.log(JSON.stringify({ row_count: rows.length }));
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main(process.argv.slice(2));
}

File: scripts/build_csv.py
"""Deterministic CSV builder (shared core) for the estimator plugin.

Reads a JSON array of Jira ticket records and writes/merges them into the
global history CSV with delimiter-aware escaping and dedup-by-key.
"""

import csv
import json
import os
import re
import sys

HEADERS = [
    "Key (key)",
    "Parent (parent_key)",
    "Issue Type (issue_type)",
    "Title (summary)",
    "Description (description)",
    "Components (components)",
    "Labels (labels)",
    "Story Points (story_points)",
    "Time Spent [s] (time_spent_seconds)",
    "Resolution Date (resolution_date)",
    "Comments (comments)",
    "Design Link (design_link)",
    "Estimate Basis (estimate_basis)",
]

FIELD_KEYS = [re.search(r"\(([^)]+)\)\s*$", h).group(1) for h in HEADERS]

MULTIVALUE_SEP = ", "
COMMENT_SEP = "\n\n"


def _str(value):
    return "" if value is None else str(value)


def _join(values, sep):
    if not values:
        return ""
    if isinstance(values, (list, tuple)):
        return sep.join(_str(v) for v in values)
    return _str(values)


def record_to_row(rec):
    return [
        _str(rec.get("key")),
        _str(rec.get("parent_key")),
        _str(rec.get("issue_type")),
        _str(rec.get("summary")),
        _str(rec.get("description")),
        _join(rec.get("components"), MULTIVALUE_SEP),
        _join(rec.get("labels"), MULTIVALUE_SEP),
        _str(rec.get("story_points")),
        _str(rec.get("time_spent_seconds")),
        _str(rec.get("resolution_date")),
        _join(rec.get("comments"), COMMENT_SEP),
        _str(rec.get("design_link")),
        _str(rec.get("estimate_basis")),
    ]


def build_rows(existing_rows, records, mode):
    if mode not in ("overwrite", "append"):
        raise ValueError("mode must be 'overwrite' or 'append', got: %r" % mode)
    merged = {}
    if mode == "append" and existing_rows:
        for row in existing_rows:
            if row:
                merged[row[0]] = row
    for rec in records:
        row = record_to_row(rec)
        merged[row[0]] = row
    return list(merged.values())


def _normalize(row):
    """Pad legacy rows that predate a column addition out to the header width."""
    if len(row) < len(HEADERS):
        return row + [""] * (len(HEADERS) - len(row))
    return row


def read_existing(path):
    if not os.path.exists(path):
        return []
    with open(path, newline="", encoding="utf-8") as f:
        rows = list(csv.reader(f, delimiter=";"))
    return [_normalize(r) for r in rows[1:]] if rows else []


def write_csv(path, rows):
    os.makedirs(os.path.dirname(os.path.abspath(path)), exist_ok=True)
    with open(path, "w", newline="", encoding="utf-8") as f:
        writer = csv.writer(
            f, delimiter=";", quoting=csv.QUOTE_MINIMAL, lineterminator="\n"
        )
        writer.writerow(HEADERS)
        writer.writerows(rows)


def main(argv):
    if len(argv) != 4:
        sys.exit("Usage: build_csv.py <input.json> <output.csv> <overwrite|append>")
    input_json, output_csv, mode = argv[1], argv[2], argv[3]
    with open(input_json, encoding="utf-8") as f:
        records = json.load(f)
    existing = read_existing(output_csv) if mode == "append" else []
    rows = build_rows(existing, records, mode)
    write_csv(output_csv, rows)
    print("Wrote %d rows to %s" % (len(rows), output_csv))
    print(json.dumps({"row_count": len(rows)}, separators=(",", ":")))


if __name__ == "__main__":
    main(sys.argv)
